src/c_api/c_api.cc

```diff
@@ -825,74 +825,67 @@ XGB_DLL int XGBoosterPredictFromDMatrix(BoosterHandle handle,
-                        xgboost::bst_ulong *out_dim, const float **out_result) {
+void InplacePredictImpl(std::shared_ptr<DMatrix> p_m, char const *c_json_config, Learner *learner,
+                        xgboost::bst_ulong const **out_shape, xgboost::bst_ulong *out_dim,
+                        const float **out_result) {
   auto config = Json::Load(StringView{c_json_config});
   CHECK_EQ(get<Integer const>(config["cache_id"]), 0) << "Cache ID is not supported yet";
 
-  HostDeviceVector<float>* p_predt { nullptr };
+  HostDeviceVector<float> *p_predt{nullptr};
   auto type = PredictionType(RequiredArg<Integer>(config, "type", __func__));
   float missing = GetMissing(config);
-  learner->InplacePredict(x, p_m, type, missing, &p_predt,
+  learner->InplacePredict(p_m, type, missing, &p_predt,
                           RequiredArg<Integer>(config, "iteration_begin", __func__),
                           RequiredArg<Integer>(config, "iteration_end", __func__));
   CHECK(p_predt);
   auto &shape = learner->GetThreadLocal().prediction_shape;
-  auto chunksize = n_rows == 0 ? 0 : p_predt->Size() / n_rows;
+  auto const &info = p_m->Info();
+  auto n_samples = info.num_row_;
+  auto n_features = info.num_col_;
+  auto chunksize = n_samples == 0 ? 0 : p_predt->Size() / n_samples;
   bool strict_shape = RequiredArg<Boolean>(config, "strict_shape", __func__);
-  CalcPredictShape(strict_shape, type, n_rows, n_cols, chunksize, learner->Groups(),
+  CalcPredictShape(strict_shape, type, n_samples, n_features, chunksize, learner->Groups(),
                    learner->BoostedRounds(), &shape, out_dim);
   *out_result = dmlc::BeginPtr(p_predt->HostVector());
   *out_shape = dmlc::BeginPtr(shape);
 }
 
 // A hidden API as cache id is not being supported yet.
-XGB_DLL int XGBoosterPredictFromDense(BoosterHandle handle,
-                                      char const *array_interface,
-                                      char const *c_json_config,
-                                      DMatrixHandle m,
+XGB_DLL int XGBoosterPredictFromDense(BoosterHandle handle, char const *array_interface,
+                                      char const *c_json_config, DMatrixHandle m,
                                       xgboost::bst_ulong const **out_shape,
-                                      xgboost::bst_ulong *out_dim,
-                                      const float **out_result) {
+                                      xgboost::bst_ulong *out_dim, const float **out_result) {
   API_BEGIN();
   CHECK_HANDLE();
-  std::shared_ptr<xgboost::data::ArrayAdapter> x{
-      new xgboost::data::ArrayAdapter(StringView{array_interface})};
-  std::shared_ptr<DMatrix> p_m {nullptr};
-  if (m) {
+  std::shared_ptr<DMatrix> p_m{nullptr};
+  if (!m) {
+    p_m.reset(new data::DMatrixProxy);
+  } else {
     p_m = *static_cast<std::shared_ptr<DMatrix> *>(m);
   }
+  dynamic_cast<data::DMatrixProxy *>(p_m.get())->SetArrayData(array_interface);
```

Comment:
How do we ensure that `m` is actually an instance of `DMatrixProxy`? The proxy is a subclass of `DMatrix`, so we could get `m` that's `DMatrix` but not `DMatrixProxy`.

Reply:
Added a check in this function and in predictor.